pkg/apis/clusterlink.net/v1alpha1/instance_types.go

```diff
@@ -0,0 +1,126 @@
+	GoType DataplaneType = "go"
+	// EnvoyType indicates that the dataplane type is Envoy.
+	EnvoyType DataplaneType = "envoy"
+)
+
+// InstanceSpec defines the desired state of a ClusterLink instance.
+type InstanceSpec struct {
+	DataPlane DataPlaneSpec `json:"dataplane,omitempty"`
+	Ingress   IngressSpec   `json:"ingress,omitempty"`
+
+	// +kubebuilder:validation:Enum=trace;debug;info;warning;error;fatal
+	// +kubebuilder:default=info
+	// LogLevel define the ClusterLink components log level.
+	LogLevel string `json:"logLevel,omitempty"`
+
+	// +kubebuilder:default="ghcr.io/clusterlink-net"
+	// ContainerRegistry is the container registry to pull the ClusterLink project images.
+	ContainerRegistry string `json:"containerRegistry,omitempty"`
+	// +kubebuilder:default="latest"
+	// ImageTag represents the version of the ClusterLink project images.
+	ImageTag string `json:"imageTag,omitempty"`
+}
+
+// InstanceStatus defines the observed state of ClusterlLink Instance.
+type InstanceStatus struct {
+	Controlplane ComponentStatus `json:"controlplane,omitempty"`
+	Dataplane    ComponentStatus `json:"dataplane,omitempty"`
+}
+
+// ComponentStatus defines the status of component in ClusterLink.
+type ComponentStatus struct {
+	// Conditions contain the status conditions.
+	Conditions map[string]metav1.Condition `json:"conditions,omitempty"`
+}
+
+// DataPlaneSpec defines the desired state of the dataplane components in ClusterLink.
+type DataPlaneSpec struct {
+	// +kubebuilder:validation:Enum=envoy;go
+	// +kubebuilder:default=envoy
+	// Type represents the dataplane type. Supports values "go" and "envoy".
+	Type DataplaneType `json:"type,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=10
+	// +kubebuilder:default=1
+	// Replicas represents the number of dataplane replicas.
+	Replicas int `json:"replicas,omitempty"`
+}
+
+// IngressSpec defines the type of the ingress component in ClusterLink.
+type IngressSpec struct {
+	// +kubebuilder:validation:Enum=none;LoadBalancer;NodePort
+	// +kubebuilder:default=none
+	// Type represents the type of service used to expose the ClusterLink deployment.
+	// Supported values: "LoadBalancer","NodePort", "none". The service name will be "cl-external-svc".
+	Type IngressType `json:"type,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// Instance is the Schema for the ClusterLink instance API used for deploying a ClusterLink instance.
+type Instance struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   InstanceSpec   `json:"spec,omitempty"`
+	Status InstanceStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// InstanceList represents a list of Instance objects.
+type InstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Instance `json:"items"`
+}
+
+func init() {
```

Comment:
question: is this where kubebuilder places the `init()` or is it copied from a different file?
I was thinking it belonged in the `groupversion_info.go` file...

Reply:
done by the kubebuilder, I just added the instance structs
https://kubebuilder.io/cronjob-tutorial/new-api#:~:text=Finally%2C%20we%20add%20the%20Go%20types%20to%20the%20API%20group.%20This%20allows%20us%20to%20add%20the%20types%20in%20this%20API%20group%20to%20any%20Scheme.